leap/multistep/__init__.py

```diff
@@ -385,35 +423,149 @@ def rk_bootstrap(self, cb):
-                                         in enumerate(coeffs))
+        # Assign the value of the new state.
+        cb(self.state, est_vars[0])
+
+# }}}
 
-                if self.state_filter is not None:
-                    stage = self.state_filter(stage)
 
-                cb(rhss[stage_num], self.eval_rhs(self.t + c * self.dt, stage))
+# {{{ am method
 
-        # Merge the values of the RHSs.
-        rk_comb = sum(coeff * rhss[j] for j, coeff in enumerate(rk_coeffs))
 
-        state_est = self.state + self.dt * rk_comb
+class AdamsMoultonMethodBuilder(AdamsMethodBuilder):
+    def generate_primary(self, cb):
+
+        from pytools import UniqueNameGenerator
+        name_gen = UniqueNameGenerator()
+        rhs_next_var = var("rhs_next_var")
+        rhs_var_to_unknown = {}
+        unkvar = cb.fresh_var("unk")
+        rhs_var_to_unknown[rhs_next_var] = unkvar
+
+        # In implicit mode, the vector of times
+        # passed to adams_integration must
+        # include the *next* point in time.
+        time_data, relv_times, \
+                dt_factor, t_end = self.set_up_time_data(cb, self.t + self.dt)
+
+        # Implicit setup - rhs_next_var is an unknown, needs implicit solve.
+        equations = []
+        unknowns = set()
+        knowns = set()
+
+        unknowns.add(rhs_next_var)
+
+        # Create RHS vector for Adams setup,
+        # including RHS value to be implicitly
+        # solved for
+        rhss = self.history + [rhs_next_var]
+
+        # Set up the actual Adams-Moulton step.
+        am_sum = emit_adams_integration(
+                        cb, name_gen,
+                        self.function_family,
+                        relv_times, rhss,
+                        0, t_end)
+
+        state_est = self.state + dt_factor * am_sum
+
+        # Build the implicit solve expression.
+        from dagrt.expression import collapse_constants
+        from pymbolic.mapper.distributor import DistributeMapper as DistMap
+        solve_expression = collapse_constants(
+                rhs_next_var - self.eval_rhs(self.t + self.dt,
+                                                 DistMap()(state_est)),
+                list(unknowns) + [self.state],
+                cb.assign, cb.fresh_var)
+        equations.append(solve_expression)
+
+        # {{{ emit solve if possible
+
+        if unknowns and len(unknowns) == len(equations):
+            from leap.implicit import generate_solve
+            generate_solve(cb, unknowns, equations, rhs_var_to_unknown, self.state)
+        elif not unknowns:
+            raise ValueError("Adams-Moulton implicit timestep has no unknowns")
+        elif len(unknowns) > len(equations):
+            raise ValueError("Adams-Moulton implicit timestep has more unknowns "
+                    "than equations")
+        elif len(unknowns) < len(equations):
+            raise ValueError("Adams-Moulton implicit timestep has more equations "
+                    "than unknowns")
```

Comment:
Is there an `else` to this last condition? If not:
```suggestion
                    "than unknowns")
        else:
            assert False
```

Reply:
Addressed in 61277d8.